lib/store.js

```diff
@@ -0,0 +1,190 @@
+const NamespaceHandlerHmac = require('./namespace-handler-hmac');
+
+const logger = bedrock.loggers.get('app');
+
+class Store {
+  constructor() {
+    const handlerWebKey = new NamespaceHandlerWebKey();
+    this.handlers = {
+      HS: new NamespaceHandlerHmac(),
+      RS: handlerWebKey
+      // TODO: Add ECDSA support
+      // ES: handlerWebKey
+    };
+  }
+
+  getNamespace(namespace, callback) {
+    database.collections.jwtKeyStore.findOne(
+      {id: database.hash(namespace)}, {namespace: true}, (err, record) => {
+        if(err) {
+          logger.error('Mongo error when trying to find key state.', err);
+          return callback(new BedrockError(
+            'Failed to find key state due to internal error.',
+            'InternalError',
+            {namespace: namespace, httpStatusCode: 500, 'public': true}, err));
+        }
+        if(!record) {
+          return callback(new BedrockError(
+            'Key state not found.',
+            'NotFound',
+            {namespace: namespace, httpStatusCode: 404, 'public': true}));
+        }
+        callback(null, record.namespace);
+      });
+  }
+
+  /**
+   * Initialize a namespaced keystore.
+   *
+   * @param namespace the namespace to initialize.
+   * @param options the options to use:
+   *   id the ID for the namespace.
+   *   algorithm the JWT signing algorithm.
+   *   clockToleranceInSecs the acceptable clock skew.
+   *   tokenTtlInSecs the JWT token TTL.
+   *   [key] the key options:
+   *     id the key ID.
+   *
+   * @param callback(err) called once the operation completes.
+   */
+  provision(options, callback) {
+    const self = this;
+    let handler;
+    try {
+      handler = self._getNamespaceHandler(options.algorithm);
+    } catch(e) {
+      return callback(new BedrockError(
+        'Unsupported algorithm.', 'NotFound',
+        {public: false, namespace: options}));
+    }
+    async.auto({
+      createState: handler.createState.bind(handler, options),
+      storeState: ['createState', (callback, results) => {
+        const now = Date.now();
+        const record = {
+          id: database.hash(options.id),
+          meta: {
+            created: now,
+            updated: now
+          },
+          namespace: {
+            id: options.id,
+            algorithm: options.algorithm,
+            clockToleranceInSecs: options.clockToleranceInSecs,
+            tokenTtlInSecs: options.tokenTtlInSecs,
+            state: results.createState
+          }
+        };
+        database.collections.jwtKeyStore.insert(
+          record, database.writeOptions, (err, result) => {
+            if(err) {
+              if(database.isDuplicateError(err)) {
+                // namespace is already provisioned
+                // TODO: ensure record in database matches given options
+                return callback();
+              }
+              logger.error('Mongo error when trying to insert key state.', err);
```

Comment:
`logger.error('[jwt-mongodb] Mongo error when trying to insert namespace.', err);`